cadence/contracts/bridge/FlowEVMBridgeNFTEscrow.cdc

```diff
@@ -0,0 +1,244 @@
+                return locker.borrowNFT(id)
+            }
+        }
+        return nil
+    }
+
+    /// Returns whether an NFT with the given ID is locked
+    ///
+    /// @param id ID of the NFT to check
+    ///
+    /// @returns True if the NFT is locked, false otherwise
+    ///
+    access(all)
+    view fun isLocked(type: Type, id: UInt64): Bool {
+        return self.borrowLockedNFT(type: type, id: id) != nil
+    }
+
+    access(all)
+    view fun getLockedCadenceID(type: Type, evmID: UInt256): UInt64? {
+        if let lockerPath = FlowEVMBridgeUtils.deriveEscrowStoragePath(fromType: type) {
+            if let locker = self.account.storage.borrow<&Locker>(from: lockerPath) {
+                return locker.getCadenceID(from: evmID)
+            }
+        }
+        return nil
+    }
+
+    /**********************
+        Bridge Methods
+    ***********************/
+
+    access(account)
+    fun initializeEscrow(forType: Type, erc721Address: EVM.EVMAddress) {
+        let lockerPath = FlowEVMBridgeUtils.deriveEscrowStoragePath(fromType: forType)
+            ?? panic("Problem deriving locker path")
+        if self.account.storage.type(at: lockerPath) != nil {
+            return
+        }
+        let locker <- create Locker(lockedType: forType, erc721Address: erc721Address)
+        self.account.storage.save(<-locker, to: lockerPath)
+    }
+
+    access(account) fun lockNFT(_ nft: @{NonFungibleToken.NFT}) {
+        let lockerPath = FlowEVMBridgeUtils.deriveEscrowStoragePath(fromType: nft.getType())
+            ?? panic("Problem deriving locker path")
+        let locker = self.account.storage.borrow<&Locker>(from: lockerPath)
+            ?? panic("Locker doesn't exist")
+        locker.deposit(token: <-nft)
+    }
+
+    access(account)
+    fun unlockNFT(type: Type, id: UInt64): @{NonFungibleToken.NFT} {
+        let lockerPath = FlowEVMBridgeUtils.deriveEscrowStoragePath(fromType: type)
+            ?? panic("Problem deriving locker path")
+        let locker = self.account.storage.borrow<auth(NonFungibleToken.Withdrawable) &Locker>(from: lockerPath)
+            ?? panic("Locker doesn't exist")
+        return <- locker.withdraw(withdrawID: id)
+    }
+
+    /*********************
+            Locker
+    *********************/
+
+    /// The resource managing the locking & unlocking of NFTs via this contract's interface
+    ///
+    access(all) resource Locker : CrossVMNFT.EVMNFTCollection, NonFungibleToken.Collection {
+        /// The type of NFTs this Locker escrows
+        access(all)
+        let lockedType: Type
+        /// Corresponding ERC721 address for the locked NFTs
+        access(all)
+        let erc721Address: EVM.EVMAddress
+        /// Count of locked NFTs as lockedNFTs.length may exceed computation limits
+        access(self)
+        var lockedNFTCount: Int
+        /// Indexed on NFT UUID to prevent collisions
+        access(self)
+        let lockedNFTs: @{UInt64: {NonFungibleToken.NFT}}
```

Comment:
wouldn't this get too big over time ?

Reply:
That's a good question, and one I was concerned about initially which prompted some local benchmarking. From what I could tell, single unit insertion computation costs will certainly increase as the absolute number of stored units increases - part of the reason why I included the length value for `getLength()`. But accessing single values either by reference or removal from the mapping should still function for an arbitrary number of stored units. Do you think that reasoning is sound?

Reply:
Accessing will be ok for sure, reading cost multiplier was approx `log40(N)` when I tried last time for UInt64 keyed dictionaries. ( for C1.0 )

Writing is unpredictable a bit. but yesterday on GWG there was some discussion about increasing max compute indirectly by fees and weights, so I think not a concern for now.

Reply:
I think also `length` should be cheap, keys should be expensive.